src/main/java/uk/co/ramp/Outbreak.java

```diff
@@ -63,171 +70,71 @@ public void setContactRecords(Map<Integer, List<ContactRecord>> contactRecords)
-            int activeCases = calculateDailyStatistics(time);
-
-            if (activeCases == 0 && randomInfectionRate == 0d) {
-                LOGGER.info("There are no active cases and the random infection rate is zero.");
-                LOGGER.info("Exiting as solution is stable.");
-                return true;
-            }
-
-            for (ContactRecord contacts : todaysContacts) {
-                evaluateContact(time, contacts);
-            }
+    void runContactData(int timeLimit, double randomInfectionRate) {
+        int lastContact = eventList.getMap().keySet().stream().max(Comparator.naturalOrder()).orElseThrow();
 
+        if (lastContact > timeLimit) {
+            LOGGER.info("timeLimit it lower than time of last contact event");
+            LOGGER.info("Not all contact data will be used");
         }
-        return false;
-    }
-
-
-    void evaluateContact(int time, ContactRecord contacts) {
-        Case potentialSpreader = population.get(contacts.to());
-        Case victim = population.get(contacts.from());
 
-        boolean conditionA = potentialSpreader.alertStatus() != NONE || victim.alertStatus() != NONE;
-        boolean conditionB = contacts.weight() < diseaseProperties.exposureThreshold();
-
-
-        if (conditionA && conditionB) {
-            // TODO: Apply behavioural logic here. Use compliance value?
-            LOGGER.trace("spreader: {}   victim: {}   weight: {} ", potentialSpreader.alertStatus(), victim.alertStatus(), contacts.weight());
-            LOGGER.debug("Skipping contact due to threshold");
-            return;
-        }
-
-
-        if (potentialSpreader.status() != victim.status()) {
-            evaluateExposures(contacts, time);
-        }
-    }
+        for (int time = 0; time <= timeLimit; time++) {
 
+            eventProcessor.setPopulation(population);
+            eventProcessor.process(time, randomInfectionRate, lastContact);
+            updateLogActiveCases(time);
 
-    void runToSteadyState(int runTime, int timeLimit) {
-        for (int time = runTime; time <= timeLimit; time++) {
-
-            // set random infection rate to zero in steady state mode
-            updatePopulationState(time, 0d);
-            calculateDailyStatistics(time);
-
-            Map<VirusStatus, Integer> compartmentCounts = PopulationGenerator.getCmptCounts(population);
-
-
-            boolean exitCondition = compartmentCounts.get(EXPOSED) == 0 &&
-                    compartmentCounts.get(EXPOSED_2) == 0 &&
-                    compartmentCounts.get(INFECTED) == 0 &&
-                    compartmentCounts.get(INFECTED_SYMP) == 0;
-
-
-            if (exitCondition) {
-                LOGGER.info("Steady state solution reached at t={} ", time);
-                LOGGER.info("Exiting early.");
-                return;
+            // stop random infections after contacts end
+            if (activeCases == 0 && lastContact < time) {
```

Comment:
I would calculate this on the fly instead of storing state in class.

Perhaps a separate private method to call from here:
```
    private int activeCases() {
        var stats = utils.getCmptCounts(population);
        return stats.get(EXPOSED) + stats.get(ASYMPTOMATIC) + stats.get(PRESYMPTOMATIC) + stats.get(SYMPTOMATIC) + stats.get(SEVERELY_SYMPTOMATIC);
    }
```

Reply:
Fair point